R/load_cell_graphs.R

```diff
@@ -214,6 +262,142 @@ LoadCellGraphs.CellGraphAssay <- function (
+      is.character(cells) &&
+      (length(cells) > 0)
+  )
+  stopifnot(
+    "'cells' must be present in 'object'" =
+      all(cells %in% colnames(object))
+  )
+  load_as <- match.arg(load_as, choices = c("bipartite", "Anode", "linegraph"))
+
+  # Check if cells are already loaded
+  loaded_graphs <- !sapply(slot(object, name = "cellgraphs")[cells], is.null)
+  if ((sum(!loaded_graphs) == 0) & !force) {
+    if (verbose && check_global_verbosity())
+      cli_alert_info("All cells are alredy loaded. Returning CellGraphAssay unmodified.")
+    return(object)
+  } else {
+    if (force) {
+      slot(object, "cellgraphs")[cells] <- rep(list(NULL), length(cells)) %>% setNames(nm = cells)
+      loaded_graphs <- !sapply(slot(object, name = "cellgraphs")[cells], is.null)
+    }
+    cells_to_load <- setdiff(cells, cells[loaded_graphs])
+    if (verbose && check_global_verbosity() & (length(cells_to_load) < length(cells)))
+      cli_alert_info(glue("{length(cells) - length(cells_to_load)} CellGraphs loaded.",
+                          " Loading remaining {length(cells_to_load)} CellGraphs."))
+  }
+
+  # Find where components are located
+  fs_map_unnested_filtered <- slot(object, name = "fs_map") %>%
+    tidyr::unnest(cols = "id_map") %>%
+    filter(current_id %in% cells)
+  fs_map_nested_filtered <- fs_map_unnested_filtered %>%
+    tidyr::nest(data = c("current_id", "original_id"))
+
+  # Load cell graphs in chunks
+  cg_list_full <- lapply(seq_len(nrow(fs_map_nested_filtered)), function(i) {
+    f <- fs_map_nested_filtered$pxl_file[i]
+
+    # Unzip the edgelist parquet file to tmpdir
+    unzip(f, exdir = tempdir(), files = "edgelist.parquet")
+    unz_pq_file <- file.path(tempdir(), "edgelist.parquet")
+    pq_file <- fs::file_temp(ext = "parquet")
+
+    # The file is renamed to make sure it has a unique path
+    fs::file_move(unz_pq_file, pq_file)
+
+    # Read the edgelist in memory, but only the necessary columns
+    ar <- arrow::read_parquet(pq_file,
+                              col_select = c("upia", "upib", "marker", "component"),
+                              as_data_frame = FALSE)
+
+    if (verbose && check_global_verbosity()) {
+      cli_alert(glue("   Loading CellGraphs for {nrow(fs_map_nested_filtered$data[[i]])} cells ",
+                     "from sample {fs_map_nested_filtered$sample[i]}"))
+    }
+
+    # Split data into chunks determined by chunk_size
+    id_data_chunks <- fs_map_nested_filtered$data[[i]] %>%
+      mutate(group = ceiling(seq_len(n())/chunk_size)) %>%
+      rename(component = current_id) %>%
+      group_by(group) %>%
+      group_split()
+
+    # Read cellgraphs
+    cg_list <- pbapply::pblapply(id_data_chunks, function(id_chunk) {
+
+      # Filter edgelist data for the current chunk
+      # Note that we use original_id which are the original
+      # MPX component ids. current_id are the ids currenty used
+      # for components in object
+      edgelist_data <- ar %>%
```

Comment:
Why is the edgelist read into memory but the graph construction is performed in chunks? Isn't the edgelist itself the item that takes up a lot of RAM? 

The filtering per chunk seems like it could be intense, since the edgelist needs to be parsed once for every chunk (?)

I am not sure about this though, this is just an observation. What do you think?

Reply:
Yeah it's a bit of a tradeoff I guess. Here, we only read the necessary columns (upia, upib, marker, component) which lowers the memory usage a bit. We are also only reading 1 edgelist at the time, so if you have merged data, we only have to read the edgelist from one sample at the time. One that edgelist has been handled, we free up memory which can be used to load from the next sample. 

Given the size of our edgelists I think this is ok, but I was thinking I could implement an alternative strategy which only fetches chunks from the data stored on disk. It is just much slower to send querys to disk, but could be a good option for massive data sets. When I was playing around with some data sets, the memory usage wasn't too bad, but I didn't try any massive data sets. We could at least have an option for this. Also, we cannot parallelize the conversion from edgelist to graphs if we read directly from disk unless we first split the files somehow. But if we reformat the edgelists to use hive-style partitioning, I think the latter option would be much better!